src/TagHelperPack/MarkdownTagHelper.cs

```diff
@@ -15,29 +20,158 @@ public class MarkdownTagHelper : TagHelper
+                _allowHtml = allowHtml;
+                _normalizeIndentation = normalizeIndentation;
             }
 
             public void WriteTo(TextWriter writer, HtmlEncoder encoder)
             {
-                var pipeline = new MarkdownPipelineBuilder().DisableHtml().Build();
-                Markdig.Markdown.ToHtml(_markdown, writer, pipeline);
+                var markdown = _normalizeIndentation ? NormalizeIndentation(_content) : _content;
+
+                if (!_allowHtml)
+                {
+                    // Convert the markdown to HTML directly to the writer. This is a bit more memory efficient than
+                    // converting the markdown to HTML in memory and then writing out the HTML.
+                    var pipeline = new MarkdownPipelineBuilder().DisableHtml().Build();
+                    Markdown.ToHtml(markdown, writer, pipeline);
+                }
+                else
+                {
+                    WriteSanitizedHtml(markdown, writer);
+                }
+            }
+
+            void WriteSanitizedHtml(string markdown, TextWriter writer)
+            {
+                var pipeline = new MarkdownPipelineBuilder().Build();
+                // I wasn't able to find a way to sanitize the HTML in a "streaming" manner so
+                // we convert the markdown to HTML in memory so that we can sanitize it.
+                var html = Markdown.ToHtml(markdown, pipeline);
+                var sanitizedHtml = _allowHtml ? SanitizeHtml(html) : html;
+                writer.Write(sanitizedHtml);
+            }
+
+            /// <summary>
+            /// Strips leading white space based on shortest indented non-empty line.
+            /// </summary>
+            /// <remarks>
+            /// The reason this method exists is indented markdown is considered to be preformatted code blocks. And the
+            /// the markdown this tag helper receives might all be indented because of where the markdown tag helper is used
+            /// within the page (such as indented within a div).
+            /// </remarks>
+            static string NormalizeIndentation(string text)
+            {
+                if (text is { Length: 0 } || !text[0].IsWhitespace())
+                {
+                    return text;
+                }
+
+                var lineReader = new StringReader(text);
+                // In the case of something like this, the first line will be empty, so we need to advance to the first
+                // non-empty line.
+                // <markdown>
+                //    @someMarkdown
+                // </markdown>
+                var firstLine = lineReader.ReadLine();
+                while (firstLine is not { Length: > 0 })
+                {
+                    firstLine = lineReader.ReadLine();
+                }
+
+                if (firstLine is not { Length: > 0 })
+                {
+                    return text;
+                }
```

Comment:
Great candidate for using `span` here to avoid the interim allocations but fine to just log an issue for that.